static/posts/pattern-matching/post.md

```diff
@@ -0,0 +1,952 @@
+exiting with an error at runtime like this should be avoided. It can lead to
+system instability and can still be difficult to debug as the errors may only
+occur under certain conditions. Fixing the problem later in the development
+cycle can be more expensive as it may require additional deployments and
+validation. Ideally, this kind of problem should be caught as early as possible,
+i. e. at compile time.
+
+There's another problem, let's say our program needs to support many more
+shapes. The `if`/`else if`/`else` syntax can become a mess to read and maintain:
+
+```c
+// ...
+
+double shape_area(Shape shape) {
+  Dimensions ds = shape.dimensions;
+  if (shape.type == CIRCLE) {
+    return M_PI * ds.r * ds.r;
+  } else if (shape.type == RECTANGLE) {
+    return ds.w * ds.h;
+  } else if (shape.type == TRIANGLE) {
+    double s = (ds.a + ds.b + ds.c) / 2;
+    return sqrt(s * (s - ds.a) * (s - ds.b) * (s - ds.c));
+  } else if (shape.type == ...) {
+    return ...
+  } else if (shape.type == ...) {
+    return ...
+  } else if (shape.type == ...) {
+    return ...
+  } else if (shape.type == ...) {
+    return ...
+  } else if (shape.type == ...) {
+    return ...
+  } else if (shape.type == ...) {
+    return ...
+  } else {
+    printf("Missing handler for shape with type: %d\n", shape.type);
+    exit(1);
+  }
+}
+
+// ...
+```
+
+One way to improve the readability slightly is to just use `if` statements:
+
+```c
+// ...
+
+double shape_area(Shape shape) {
+  Dimensions ds = shape.dimensions;
+  if (shape.type == CIRCLE) {
+    return M_PI * ds.r * ds.r;
+  }
+  if (shape.type == RECTANGLE) {
+    return ds.w * ds.h;
+  }
+  if (shape.type == TRIANGLE) {
+    double s = (ds.a + ds.b + ds.c) / 2;
+    return sqrt(s * (s - ds.a) * (s - ds.b) * (s - ds.c));
+  }
+  if (shape.type == ...) {
+    return ...;
+  }
+  if (shape.type == ...) {
+    return ...;
+  }
+  if (shape.type == ...) {
+    return ...;
+  }
+  if (shape.type == ...) {
+    return ...;
+  }
+  if (shape.type == ...) {
+    return ...;
+  }
+  if (shape.type == ...) {
+    return ...;
+  }
+  printf("Missing handler for shape with type: %d\n", shape.type);
+  exit(1);
+}
+
+// ...
+```
+
+Another option is to use `switch`/`case` which may still be a bit easier to
+read:
+
+```c
+// ...
+
+double shape_area(Shape shape) {
+  Dimensions ds = shape.dimensions;
+  switch (shape.type) {
+    case CIRCLE:
+      return M_PI * ds.r * ds.r;
+    case RECTANGLE:
+      return ds.w * ds.h;
+    case TRIANGLE:
+      double s = (ds.a + ds.b + ds.c) / 2;
+      return sqrt(s * (s - ds.a) * (s - ds.b) * (s - ds.c));
+    case ...:
+      return ...;
+    default:
+      printf("Missing handler for shape with type: %d\n", shape.type);
+      exit(1);
+  }
+}
+
+// ...
+```
+
+However, I'd argue that it doesn't solve the main problem we're dealing with,
+and it gives you another way to shoot yourself in the foot: If you or someone
```

Comment:
I'd put a period after "foot." Then start the next sentence with "For example, if you..."